File: ios/App/XrayConfigBuilder.swift
import Foundation

/// Builds the intermediate Xray JSON (DNS + outbounds + routing) from a profile and the
/// app settings. SwiftyXrayKit injects the TUN inbound, so we omit inbounds here.
/// Mirrors the Android XrayConfigBuilder.
enum XrayConfigBuilder {

    static func build(_ p: VpnProfile, _ s: AppSettings) -> String {
        var root: [String: Any] = [
            "log": ["loglevel": "warning"],
            "dns": dns(s),
            "outbounds": outbounds(p, s),
            "routing": routing(s),
            "stats": [:],
        ]
        if s.enableFakeDns {
            root["fakedns"] = [["ipPool": "198.18.0.0/15", "poolSize": 65535]]
        }
        let data = (try? JSONSerialization.data(withJSONObject: root, options: [.prettyPrinted])) ?? Data()
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    // ── DNS ──────────────────────────────────────────────────────────────────

    private static func dns(_ s: AppSettings) -> [String: Any] {
        var servers: [Any] = []
        if s.enableFakeDns { servers.append("fakedns") }
        servers.append(["address": s.remoteDns, "domains": ["geosite:geolocation-!cn"]])
        servers.append(s.directDns)
        servers.append("localhost")
        return [
            "hosts": [
                "dns.google": ["8.8.8.8", "8.8.4.4"],
                "cloudflare-dns.com": ["1.1.1.1", "1.0.0.1"],
            ],
            "servers": servers,
        ]
    }

    // ── Outbounds ──────────────────────────────────────────────────────────────

    private static func outbounds(_ p: VpnProfile, _ s: AppSettings) -> [Any] {
        var list: [Any] = [proxyOutbound(p, s)]
        if p.fragmentEnabled {
            list.append([
                "tag": "fragment",
                "protocol": "freedom",
                "settings": ["domainStrategy": "AsIs", "fragment": [
                    "packets": s.fragmentPackets.isEmpty ? "tlshello" : s.fragmentPackets,
                    "length": s.fragmentLength.isEmpty ? "100-200" : s.fragmentLength,
                    "interval": s.fragmentInterval.isEmpty ? "10-20" : s.fragmentInterval,
                ]],
                "streamSettings": ["sockopt": ["tcpNoDelay": true]],
            ])
        }
        list.append(["tag": "direct", "protocol": "freedom",
                     "settings": ["domainStrategy": s.enableIpv6 ? "UseIP" : "UseIPv4"]])
        list.append(["tag": "block", "protocol": "blackhole",
                     "settings": ["response": ["type": "http"]]])
        return list
    }

    private static func proxyOutbound(_ p: VpnProfile, _ s: AppSettings) -> [String: Any] {
        var o: [String: Any] = ["tag": "proxy"]
        o["protocol"] = xrayProtocol(p.proto)
        o["settings"] = settings(p)

        let noStream: Set<ProxyProtocol> = [.hysteria2, .wireguard, .shadowsocks, .http]
        if !noStream.contains(p.proto) {
            var stream = streamSettings(p)
            if p.fragmentEnabled && p.transport != .quic {
                stream["sockopt"] = ["dialerProxy": "fragment"]
            }
            o["streamSettings"] = stream
        }

        let noMux: Set<ProxyProtocol> = [.shadowsocks, .trojan, .anytls, .hysteria2, .wireguard, .tuic, .socks5, .http]
        let useMux = p.muxEnabled && !noMux.contains(p.proto)
            && p.transport != .xhttp && p.transport != .quic
            && p.security != .reality && p.security != .xtls
        o["mux"] = useMux ? ["enabled": true, "concurrency": 8] : ["enabled": false]
        return o
    }

    private static func xrayProtocol(_ p: ProxyProtocol) -> String {
        switch p {
        case .vmess: return "vmess"
        case .vless: return "vless"
        case .shadowsocks: return "shadowsocks"
        case .trojan: return "trojan"
        case .hysteria2: return "hysteria2"
        case .wireguard: return "wireguard"
        case .socks5: return "socks"
        case .http: return "http"
        case .tuic: return "tuic"
        case .anytls: return "anytls"
        }
    }

    private static func settings(_ p: VpnProfile) -> [String: Any] {
        switch p.proto {
        case .vless:
            var user: [String: Any] = ["id": p.uuid, "encryption": "none"]
            if p.transport == .tcp && (p.security == .reality || p.security == .xtls) {
                user["flow"] = "xtls-rprx-vision"
            }
            return ["vnext": [["address": p.address, "port": p.port, "users": [user]]]]
        case .vmess:
            return ["vnext": [["address": p.address, "port": p.port, "users":
                [["id": p.uuid, "alterId": 0, "security": p.vmessSecurity.isEmpty ? "auto" : p.vmessSecurity]]]]]
        case .shadowsocks:
            return ["servers": [["address": p.address, "port": p.port, "method": p.ssMethod, "password": p.ssPassword]]]
        case .trojan:
            return ["servers": [["address": p.address, "port": p.port, "password": p.uuid]]]
        case .anytls:
            return ["address": p.address, "port": p.port, "password": p.uuid]
        case .hysteria2:
            var server: [String: Any] = ["address": p.address, "port": p.port, "password": p.hystPassword]
            if !p.sni.isEmpty { server["tlsSettings"] = ["serverName": p.sni, "allowInsecure": p.allowInsecure] }
            if !p.hystObfs.isEmpty { server["obfs"] = ["type": p.hystObfs, "password": p.hystObfsPassword] }
            return ["servers": [server]]
        case .wireguard:
            var settings: [String: Any] = [
                "secretKey": p.wgPrivateKey,
                "address": [p.address.isEmpty ? "10.0.0.2/32" : p.address],
                "peers": [[
                    "publicKey": p.wgPeerPublicKey,
                    "allowedIPs": ["0.0.0.0/0", "::/0"],
                    "endpoint": p.wgEndpoint.isEmpty ? "\(p.address.components(separatedBy: "/").first ?? p.address):\(p.port)" : p.wgEndpoint,
                ]],
                "mtu": p.wgMtu,
                "domainStrategy": "ForceIP",
            ]
            if let bytes = reservedBytes(p.wgReserved) { settings["reserved"] = bytes }
            if !p.wgPreSharedKey.isEmpty, var peers = settings["peers"] as? [[String: Any]] {
                peers[0]["preSharedKey"] = p.wgPreSharedKey
                settings["peers"] = peers
            }
            return settings
        case .socks5, .http:
            var server: [String: Any] = ["address": p.address, "port": p.port]
            if p.uuid.contains(":") {
                server["users"] = [["user": p.uuid.components(separatedBy: ":").first ?? "",
                                    "pass": p.uuid.components(separatedBy: ":").dropFirst().joined(separator: ":")]]
            }
            return ["servers": [server]]
        case .tuic:
            return ["server": p.address, "port": p.port, "uuid": p.uuid, "password": p.ssPassword,
                    "congestionControl": "bbr", "udpRelayMode": "native", "zeroRttHandshake": false]
        }
    }

    private static func streamSettings(_ p: VpnProfile) -> [String: Any] {
        var s: [String: Any] = [:]
        let network: String = {
            switch p.transport {
            case .tcp: return "tcp"; case .ws: return "ws"; case .grpc: return "grpc"
            case .xhttp: return "xhttp"; case .h2: return "http"; case .quic: return "quic"
            }
        }()
        s["network"] = network

        switch p.transport {
        case .ws:
            var ws: [String: Any] = ["path": p.path.isEmpty ? "/" : p.path]
            if !p.host.isEmpty { ws["headers"] = ["Host": p.host] }
            s["wsSettings"] = ws
        case .grpc:
            s["grpcSettings"] = ["serviceName": p.path, "multiMode": false]
        case .h2:
            var h: [String: Any] = ["path": p.path.isEmpty ? "/" : p.path]
            if !p.host.isEmpty { h["host"] = [p.host] }
            s["httpSettings"] = h
        case .xhttp:
            s["xhttpSettings"] = ["path": p.path.isEmpty ? "/" : p.path,
                                  "host": p.host.isEmpty ? (p.sni.isEmpty ? p.address : p.sni) : p.host,
                                  "mode": "stream-one"]
        case .tcp, .quic:
            break
        }

        switch p.security {
        case .tls:
            s["security"] = "tls"
            var tls: [String: Any] = ["serverName": p.sni.isEmpty ? p.address : p.sni,
                                      "allowInsecure": p.allowInsecure]
            if !p.fingerprint.isEmpty { tls["fingerprint"] = p.fingerprint }
            s["tlsSettings"] = tls
        case .reality:
            s["security"] = "reality"
            s["realitySettings"] = ["serverName": p.sni.isEmpty ? p.address : p.sni,
                                    "fingerprint": p.fingerprint.isEmpty ? "chrome" : p.fingerprint,
                                    "shortId": p.shortId, "publicKey": p.publicKey, "spiderX": ""]
        case .xtls:
            s["security"] = "tls"
            s["tlsSettings"] = ["serverName": p.sni.isEmpty ? p.address : p.sni, "allowInsecure": p.allowInsecure]
        case .none:
            s["security"] = "none"
        }
        return s
    }

    // ── Routing ────────────────────────────────────────────────────────────────

    private static func routing(_ s: AppSettings) -> [String: Any] {
        let applyDirect = s.routingMode == .ruleBased
        let applyChina = s.routingMode == .ruleBased && s.bypassChina
        var rules: [[String: Any]] = []

        if !s.customBlockedDomains.isEmpty {
            rules.append(["type": "field", "outboundTag": "block", "domain": s.customBlockedDomains])
        }
        if s.blockAds {
            rules.append(["type": "field", "outboundTag": "block", "domain": ["geosite:category-ads-all"]])
        }
        if applyDirect && !s.customDirectDomains.isEmpty {
            rules.append(["type": "field", "outboundTag": "direct", "domain": s.customDirectDomains])
        }
        if applyChina {
            rules.append(["type": "field", "outboundTag": "direct", "domain": ["geosite:cn"]])
            rules.append(["type": "field", "outboundTag": "direct", "ip": ["geoip:cn"]])
        }
        if s.routingMode != .global {
            rules.append(["type": "field", "outboundTag": "direct",
                          "ip": ["geoip:private", "127.0.0.0/8", "10.0.0.0/8",
                                 "172.16.0.0/12", "192.168.0.0/16", "169.254.0.0/16"]])
        }
        rules.append(["type": "field", "outboundTag": "proxy", "network": "tcp,udp"])

        return ["domainStrategy": s.domainStrategy.rawValue, "domainMatcher": "hybrid", "rules": rules]
    }

    private static func reservedBytes(_ raw: String) -> [Int]? {
        let parts = raw.components(separatedBy: CharacterSet(charactersIn: ", "))
            .compactMap { Int($0) }.filter { (0...255).contains($0) }
        return parts.count == 3 ? parts : nil
    }
}

File: linux/src/main/kotlin/com/palazik/vpn/service/XrayConfigBuilder.kt
package com.palazik.vpn.service

import com.palazik.vpn.data.model.*
import org.json.JSONArray
import org.json.JSONObject

object XrayConfigBuilder {

    // Same local ports the Android app uses, so docs/tooling stay consistent.
    const val SOCKS_PORT = 10808
    const val HTTP_PORT = 10809
    const val API_PORT = 10085

    fun build(profile: VpnProfile, settings: AppSettings = AppSettings()): String =
        JSONObject().apply {
            put("log",       buildLog())
            put("dns",       buildDns(settings))
            // FakeDNS pool — speeds up routing and avoids real DNS leaks
            if (settings.enableFakeDns) {
                put("fakedns", JSONArray().apply {
                    put(JSONObject().apply {
                        put("ipPool",   "198.18.0.0/15")
                        put("poolSize", 65535)
                    })
                })
            }
            put("inbounds",  buildInbounds(settings))
            put("outbounds", buildOutbounds(profile, settings))
            put("routing",   buildRouting(settings))
            put("stats",     JSONObject())
            put("policy",    buildPolicy())
            // gRPC StatsService on the api inbound — the desktop equivalent of
            // libv2ray's queryStats; polled via `xray api statsquery`.
            put("api", JSONObject().apply {
                put("tag", "api")
                put("services", JSONArray().apply { put("StatsService") })
            })
        }.toString(2)

    // ── Log ───────────────────────────────────────────────────────────────────

    private fun buildLog() = JSONObject().apply {
        put("loglevel", "warning")
    }

    // ── Inbounds ──────────────────────────────────────────────────────────────

    // Unlike Android (where libv2ray owns a TUN fd), the desktop xray binary has no
    // TUN inbound: tun2socks feeds full-device traffic into the SOCKS inbound instead.
    private fun buildInbounds(settings: AppSettings) = JSONArray().apply {
        // FakeDNS must be the first destOverride entry so sniffed connections map back
        // to their real domain before http/tls override (v2rayNG ordering).
        fun sniffing() = JSONObject().apply {
            put("enabled", true)
            put("destOverride", JSONArray().apply {
                if (settings.enableFakeDns) put("fakedns")
                put("http"); put("tls")
            })
        }
        put(JSONObject().apply {
            put("tag", "socks")
            put("port", SOCKS_PORT)
            put("listen", "127.0.0.1")
            put("protocol", "socks")
            put("settings", JSONObject().apply {
                put("auth", "noauth")
                put("udp", true)
                put("userLevel", 8)
            })
            put("sniffing", sniffing())
        })
        put(JSONObject().apply {
            put("tag", "http")
            put("port", HTTP_PORT)
            put("listen", "127.0.0.1")
            put("protocol", "http")
            put("settings", JSONObject().apply { put("userLevel", 8) })
        })
        put(JSONObject().apply {
            put("tag", "api")
            put("port", API_PORT)
            put("listen", "127.0.0.1")
            put("protocol", "dokodemo-door")
            put("settings", JSONObject().apply { put("address", "127.0.0.1") })
        })
    }

    // ── Outbounds ─────────────────────────────────────────────────────────────

    private fun buildOutbounds(profile: VpnProfile, settings: AppSettings) = JSONArray().apply {
        put(buildProxyOutbound(profile, settings))
        // Anti-DPI fragment dialer — the proxy outbound chains through this via
        // sockopt.dialerProxy when the profile opts in. Only added when needed.
        if (profile.fragmentEnabled) {
            put(JSONObject().apply {
                put("tag", "fragment")
                put("protocol", "freedom")
                put("settings", JSONObject().apply {
                    put("domainStrategy", "AsIs")
                    put("fragment", JSONObject().apply {
                        put("packets",  settings.fragmentPackets.ifBlank { "tlshello" })
                        put("length",   settings.fragmentLength.ifBlank { "100-200" })
                        put("interval", settings.fragmentInterval.ifBlank { "10-20" })
                    })
                })
                put("streamSettings", JSONObject().apply {
                    put("sockopt", JSONObject().apply { put("tcpNoDelay", true) })
                })
            })
        }
        put(JSONObject().apply {
            put("tag", "direct")
            put("protocol", "freedom")
            // UseIP allows IPv6 dialling when the user opted into IPv6; otherwise force IPv4.
            put("settings", JSONObject().apply {
                put("domainStrategy", if (settings.enableIpv6) "UseIP" else "UseIPv4")
            })
        })
        put(JSONObject().apply {
            put("tag", "block")
            put("protocol", "blackhole")
            put("settings", JSONObject().apply {
                put("response", JSONObject().apply { put("type", "http") })
            })
        })
        // Intercepts DNS queries arriving through the tunnel and resolves them via the
        // configured DNS section instead of leaking them upstream.
        put(JSONObject().apply {
            put("tag", "dns-out")
            put("protocol", "dns")
        })
    }

    private fun buildProxyOutbound(profile: VpnProfile, settings: AppSettings) = JSONObject().apply {
        put("tag", "proxy")

        when (profile.protocol) {
            Protocol.VLESS       -> buildVless(this, profile)
            Protocol.VMESS       -> buildVmess(this, profile)
            Protocol.SHADOWSOCKS -> buildShadowsocks(this, profile)
            Protocol.TROJAN      -> buildTrojan(this, profile)
            Protocol.HYSTERIA2   -> buildHysteria2(this, profile)
            Protocol.WIREGUARD   -> buildWireguard(this, profile)
            Protocol.SOCKS5      -> buildSocksOut(this, profile)
            Protocol.HTTP        -> buildHttpOut(this, profile)
            Protocol.TUIC        -> buildTuic(this, profile)
            Protocol.ANYTLS      -> buildAnyTls(this, profile)
        }

        // SS handles its own framing; streamSettings causes TLS handshake against plain SS servers
        val needsStream = profile.protocol !in listOf(
            Protocol.HYSTERIA2, Protocol.WIREGUARD, Protocol.SHADOWSOCKS, Protocol.HTTP
        )
        if (needsStream) {
            val stream = buildStreamSettings(profile)
            // Chain through the fragment dialer for anti-DPI. Only meaningful
            // for TCP-based transports; QUIC/UDP transports ignore TCP fragmentation.
            if (profile.fragmentEnabled && profile.transport != Transport.QUIC) {
                stream.put("sockopt", JSONObject().apply { put("dialerProxy", "fragment") })
            }
            put("streamSettings", stream)
        }

        // v2rayNG: mux disabled for SS, Trojan, WireGuard, Hysteria2, TUIC, SOCKS5, AnyTLS
        // Shadowsocks + mux is broken — xray-core does not support it
        // REALITY/XTLS use xtls-rprx-vision flow which is incompatible with mux
        // AnyTLS has its own session multiplexing, so xray mux must stay off.
        // Per-profile override: the user can force mux off via profile.muxEnabled.
        val useMux = profile.muxEnabled && profile.protocol !in listOf(
            Protocol.SHADOWSOCKS, Protocol.TROJAN, Protocol.ANYTLS,
            Protocol.HYSTERIA2, Protocol.WIREGUARD, Protocol.TUIC, Protocol.SOCKS5, Protocol.HTTP
        ) && profile.transport !in listOf(Transport.XHTTP, Transport.QUIC)
          && profile.security !in listOf(Security.REALITY, Security.XTLS)
        put("mux", JSONObject().apply {
            put("enabled", useMux)
            if (useMux) put("concurrency", 8)
        })
    }

    // ── Protocol builders ─────────────────────────────────────────────────────

    private fun buildVless(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "vless")
        obj.put("settings", JSONObject().apply {
            put("vnext", JSONArray().apply {
                put(JSONObject().apply {
                    put("address", p.address)
                    put("port", p.port)
                    put("users", JSONArray().apply {
                        put(JSONObject().apply {
                            put("id", p.uuid)
                            put("encryption", "none")
                            // xtls-rprx-vision flow is only valid on raw/TCP transport.
                            // Setting it on WS/gRPC/XHTTP/H2/QUIC makes xray reject the outbound,
                            // so only apply flow for REALITY/XTLS over TCP.
                            val flow = when {
                                p.transport == Transport.TCP &&
                                    (p.security == Security.REALITY || p.security == Security.XTLS) ->
                                    "xtls-rprx-vision"
                                else -> ""
                            }
                            if (flow.isNotEmpty()) put("flow", flow)
                        })
                    })
                })
            })
        })
    }

    private fun buildVmess(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "vmess")
        obj.put("settings", JSONObject().apply {
            put("vnext", JSONArray().apply {
                put(JSONObject().apply {
                    put("address", p.address)
                    put("port", p.port)
                    put("users", JSONArray().apply {
                        put(JSONObject().apply {
                            put("id", p.uuid)
                            put("alterId", 0)
                            put("security", p.vmessSecurity.ifBlank { "auto" })
                        })
                    })
                })
            })
        })
    }

    private fun buildShadowsocks(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "shadowsocks")
        obj.put("settings", JSONObject().apply {
            put("servers", JSONArray().apply {
                put(JSONObject().apply {
                    put("address",  p.address)
                    put("port",     p.port)
                    put("method",   p.ssMethod)
                    put("password", p.ssPassword)
                })
            })
        })
    }

    private fun buildTrojan(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "trojan")
        obj.put("settings", JSONObject().apply {
            put("servers", JSONArray().apply {
                put(JSONObject().apply {
                    put("address",  p.address)
                    put("port",     p.port)
                    put("password", p.uuid)
                })
            })
        })
    }

    private fun buildHysteria2(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "hysteria2")
        obj.put("settings", JSONObject().apply {
            put("servers", JSONArray().apply {
                put(JSONObject().apply {
                    put("address",  p.address)
                    put("port",     p.port)
                    put("password", p.hystPassword)
                    if (p.sni.isNotEmpty()) {
                        put("tlsSettings", JSONObject().apply {
                            put("serverName", p.sni)
                            put("allowInsecure", p.allowInsecure)
                        })
                    }
                    if (p.hystObfs.isNotEmpty()) {
                        put("obfs", JSONObject().apply {
                            put("type",     p.hystObfs)
                            put("password", p.hystObfsPassword)
                        })
                    }
                })
            })
        })
    }

    private fun buildWireguard(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "wireguard")
        obj.put("settings", JSONObject().apply {
            put("secretKey", p.wgPrivateKey)
            put("address", JSONArray().apply { put(p.address.ifEmpty { "10.0.0.2/32" }) })
            put("peers", JSONArray().apply {
                put(JSONObject().apply {
                    put("publicKey",  p.wgPeerPublicKey)
                    put("allowedIPs", JSONArray().apply { put("0.0.0.0/0"); put("::/0") })
                    // The endpoint is the peer's public host:port. p.address is the LOCAL
                    // tunnel address (often a CIDR like 10.0.0.2/32), so strip the CIDR in
                    // the fallback to avoid an invalid endpoint like "10.0.0.2/32:51820".
                    put("endpoint",   p.wgEndpoint.ifEmpty { "${p.address.substringBefore("/")}:${p.port}" })
                    if (p.wgPreSharedKey.isNotEmpty()) put("preSharedKey", p.wgPreSharedKey)
                })
            })
            put("mtu", p.wgMtu)
            // Cloudflare WARP "reserved" — 3 bytes prepended to each WG packet.
            parseReserved(p.wgReserved)?.let { bytes ->
                put("reserved", JSONArray().apply { bytes.forEach { put(it) } })
            }
            // WireGuard outbound only accepts Force* strategies (NOT UseIP/UseIPv4 like
            // freedom/routing) — "UseIPv4" makes xray reject the whole config.
            if (p.wgDns.isNotEmpty()) put("domainStrategy", "ForceIP")
        })
    }

    /** Parse a WARP "reserved" value ("a,b,c") into exactly three ints, or null. */
    private fun parseReserved(raw: String): List<Int>? {
        if (raw.isBlank()) return null
        val parts = raw.split(",", " ").map { it.trim() }.filter { it.isNotEmpty() }
        if (parts.size != 3) return null
        val ints = parts.mapNotNull { it.toIntOrNull()?.takeIf { n -> n in 0..255 } }
        return ints.takeIf { it.size == 3 }
    }

    private fun buildSocksOut(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "socks")
        obj.put("settings", JSONObject().apply {
            put("servers", JSONArray().apply {
                put(JSONObject().apply {
                    put("address", p.address)
                    put("port",    p.port)
                    if (p.uuid.isNotEmpty() && p.uuid.contains(":")) {
                        put("users", JSONArray().apply {
                            put(JSONObject().apply {
                                put("user", p.uuid.substringBefore(":"))
                                put("pass", p.uuid.substringAfter(":"))
                            })
                        })
                    }
                })
            })
        })
    }

    private fun buildHttpOut(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "http")
        obj.put("settings", JSONObject().apply {
            put("servers", JSONArray().apply {
                put(JSONObject().apply {
                    put("address", p.address)
                    put("port", p.port)
                    if (p.uuid.isNotEmpty() && p.uuid.contains(":")) {
                        put("users", JSONArray().apply {
                            put(JSONObject().apply {
                                put("user", p.uuid.substringBefore(":"))
                                put("pass", p.uuid.substringAfter(":"))
                            })
                        })
                    }
                })
            })
        })
    }

    private fun buildTuic(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "tuic")
        obj.put("settings", JSONObject().apply {
            put("server",            p.address)
            put("port",              p.port)
            put("uuid",              p.uuid)
            put("password",          p.ssPassword)
            put("congestionControl", "bbr")
            put("udpRelayMode",      "native")
            put("zeroRttHandshake",  false)
        })
    }

    private fun buildAnyTls(obj: JSONObject, p: VpnProfile) {
        obj.put("protocol", "anytls")
        obj.put("settings", JSONObject().apply {
            put("address",  p.address)
            put("port",     p.port)
            put("password", p.uuid)
        })
    }

    // ── Stream settings ───────────────────────────────────────────────────────

    private fun buildStreamSettings(p: VpnProfile) = JSONObject().apply {
        val network = when (p.transport) {
            Transport.TCP   -> "tcp"
            Transport.WS    -> "ws"
            Transport.GRPC  -> "grpc"
            Transport.XHTTP -> "xhttp"
            Transport.H2    -> "http"
            Transport.QUIC  -> "quic"
        }
        put("network", network)

        when (p.transport) {
            Transport.XHTTP -> put("xhttpSettings", JSONObject().apply {
                put("path", p.path.ifEmpty { "/" })
                put("host", p.host.ifEmpty { p.sni.ifEmpty { p.address } })
                put("mode", "stream-one")
            })
            Transport.WS -> put("wsSettings", JSONObject().apply {
                put("path", p.path.ifEmpty { "/" })
                if (p.host.isNotEmpty()) put("headers", JSONObject().apply { put("Host", p.host) })
            })
            Transport.GRPC -> put("grpcSettings", JSONObject().apply {
                put("serviceName", p.path.ifEmpty { "" })
                put("multiMode", false)
            })
            Transport.H2 -> put("httpSettings", JSONObject().apply {
                put("path", p.path.ifEmpty { "/" })
                if (p.host.isNotEmpty()) put("host", JSONArray().apply { put(p.host) })
            })
            Transport.TCP -> {
                if (p.host.isNotEmpty()) {
                    put("tcpSettings", JSONObject().apply {
                        put("header", JSONObject().apply {
                            put("type", "http")
                            put("request", JSONObject().apply {
                                put("version", "1.1")
                                put("method",  "GET")
                                put("path",    JSONArray().apply { put(p.path.ifEmpty { "/" }) })
                                put("headers", JSONObject().apply {
                                    put("Host",       JSONArray().apply { put(p.host) })
                                    put("User-Agent", JSONArray().apply { put("Mozilla/5.0") })
                                })
                            })
                        })
                    })
                }
            }
            else -> {}
        }

        when (p.security) {
            Security.TLS -> {
                put("security", "tls")
                put("tlsSettings", JSONObject().apply {
                    put("serverName",    p.sni.ifEmpty { p.address })
                    put("allowInsecure", p.allowInsecure)
                    if (p.fingerprint.isNotEmpty()) put("fingerprint", p.fingerprint)
                    put("alpn", JSONArray().apply {
                        if (p.transport in listOf(Transport.GRPC, Transport.H2)) put("h2")
                        put("http/1.1")
                    })
                })
            }
            Security.REALITY -> {
                put("security", "reality")
                put("realitySettings", JSONObject().apply {
                    put("serverName",  p.sni.ifEmpty { p.address })
                    put("fingerprint", p.fingerprint.ifEmpty { "chrome" })
                    put("shortId",     p.shortId)
                    put("publicKey",   p.publicKey)
                    put("spiderX",     "")
                })
            }
            Security.XTLS -> {
                put("security", "tls")
                put("tlsSettings", JSONObject().apply {
                    put("serverName",    p.sni.ifEmpty { p.address })
                    put("allowInsecure", p.allowInsecure)
                    if (p.fingerprint.isNotEmpty()) put("fingerprint", p.fingerprint)
                })
            }
            Security.NONE -> put("security", "none")
        }
    }

    // ── DNS ───────────────────────────────────────────────────────────────────

    // v2rayNG CoreConfigManager.getDns():
    // - hosts map pre-resolves known DoH provider domains to their IPs
    //   to prevent DNS loop (xray intercepts DNS → tries to resolve DoH domain
    //   → DNS intercept again → infinite loop)
    // - remote DNS (1.1.1.1 DoH) for proxied traffic
    // - direct DNS (223.5.5.5) for local/cn traffic
    private fun buildDns(settings: AppSettings) = JSONObject().apply {
        put("tag", "dns-in")
        put("hosts", JSONObject().apply {
            put("dns.google",         JSONArray().apply { put("8.8.8.8"); put("8.8.4.4") })
            put("one.one.one.one",    JSONArray().apply { put("1.1.1.1"); put("1.0.0.1") })
            put("cloudflare-dns.com", JSONArray().apply { put("1.1.1.1"); put("1.0.0.1") })
            put("dns.cloudflare.com", JSONArray().apply { put("1.1.1.1"); put("1.0.0.1") })
            put("dns.alidns.com",     JSONArray().apply { put("223.5.5.5"); put("223.6.6.6") })
            put("clients4.google.com", "clients.google.com")
        })
        put("servers", JSONArray().apply {
            // FakeDNS server must come first so A/AAAA queries get a fake IP from the pool.
            if (settings.enableFakeDns) put("fakedns")
            put(JSONObject().apply {
                put("address", settings.remoteDns)
                put("domains", JSONArray().apply { put("geosite:geolocation-!cn") })
            })
            put(settings.directDns)
            put("localhost")
        })
    }

    // ── Policy ────────────────────────────────────────────────────────────────

    private fun buildPolicy() = JSONObject().apply {
        put("levels", JSONObject().apply {
            put("8", JSONObject().apply {
                put("handshake",    4)
                put("connIdle",     300)
                put("uplinkOnly",   1)
                put("downlinkOnly", 1)
            })
        })
        put("system", JSONObject().apply {
            put("statsOutboundUplink",   true)
            put("statsOutboundDownlink", true)
        })
    }

    // ── Routing ───────────────────────────────────────────────────────────────

    // The DNS intercept rule is critical in TUN mode — without it DNS packets coming
    // through tun2socks go into the proxy, get resolved upstream, return through the
    // tunnel and get intercepted again → loop.
    private fun buildRouting(settings: AppSettings) = JSONObject().apply {
        put("domainStrategy", settings.domainStrategy.name)
        put("domainMatcher",  "hybrid")
        // GLOBAL forces everything (except LAN/DNS) through the proxy; BYPASS_LAN ignores
        // China/custom-direct bypass. Only RULE_BASED applies the full direct/bypass rules.
        val applyDirectRules = settings.routingMode == RoutingMode.RULE_BASED
        val applyChina       = settings.routingMode == RoutingMode.RULE_BASED && settings.bypassChina
        put("rules", JSONArray().apply {

            // Rule 0: gRPC stats API traffic → api outbound (MUST come before everything)
            put(JSONObject().apply {
                put("type", "field")
                put("inboundTag", JSONArray().apply { put("api") })
                put("outboundTag", "api")
            })

            // Rule 1: DNS arriving through the tunnel → dns-out (MUST be first real rule)
            put(JSONObject().apply {
                put("type", "field")
                put("inboundTag", JSONArray().apply { put("socks") })
                put("outboundTag", "dns-out")
                put("port", "53")
            })

            // User-defined blocked domains (always honoured — blocking is not a "direct" rule)
            if (settings.customBlockedDomains.isNotEmpty()) {
                put(JSONObject().apply {
                    put("type", "field")
                    put("outboundTag", "block")
                    put("domain", JSONArray().apply { settings.customBlockedDomains.forEach { put(it) } })
                })
            }

            // Block ads (optional, always honoured)
            if (settings.blockAds) {
                put(JSONObject().apply {
                    put("type", "field")
                    put("outboundTag", "block")
                    put("domain", JSONArray().apply { put("geosite:category-ads-all") })
                })
            }

            // User-defined direct domains
            if (applyDirectRules && settings.customDirectDomains.isNotEmpty()) {
                put(JSONObject().apply {
                    put("type", "field")
                    put("outboundTag", "direct")
                    put("domain", JSONArray().apply { settings.customDirectDomains.forEach { put(it) } })
                })
            }

            // Bypass China (optional) — route mainland domains & IPs direct
            if (applyChina) {
                put(JSONObject().apply {
                    put("type", "field")
                    put("outboundTag", "direct")
                    put("domain", JSONArray().apply { put("geosite:cn") })
                })
                put(JSONObject().apply {
                    put("type", "field")
                    put("outboundTag", "direct")
                    put("ip", JSONArray().apply { put("geoip:cn") })
                })
            }

            // Private/LAN IPs → direct. GLOBAL forces everything (LAN included) through the
            // proxy, so the bypass only applies to RULE_BASED and BYPASS_LAN.
            if (settings.routingMode != RoutingMode.GLOBAL) {
                put(JSONObject().apply {
                    put("type", "field")
                    put("outboundTag", "direct")
                    put("ip", JSONArray().apply {
                        put("geoip:private")
                        put("127.0.0.0/8")
                        put("10.0.0.0/8")
                        put("172.16.0.0/12")
                        put("192.168.0.0/16")
                        put("169.254.0.0/16")
                        put("224.0.0.0/4")
                    })
                })
            }

            // All other traffic → proxy
            put(JSONObject().apply {
                put("type", "field")
                put("outboundTag", "proxy")
                put("network", "tcp,udp")
            })
        })
    }
}
